### PR TITLE
Add VJP-delta activation steering

### PR DESCRIPTION
**Pull request by wassname**
Adds VJP-delta activation steering, a simplified version of Anthropic’s [Jacobian lens](https://transformer-circuits.pub/2026/workspace/) adapted to contrastive prompt pairs. Based on [vjp-steering](https://github.com/wassname/vjp-steering).

Fits steering vectors with gradients during `steer()`, then uses the existing additive intervention during generation. Includes a usage notebook.

Checks: existing VJP/lifecycle tests and notebook pass after review fixes. Full suite and CUDA smoke passed before those fixes.

### CODE REVIEW

File: examples/notebooks/algorithms/vjp_delta.ipynb [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+   },
+   "outputs": [],
+   "source": [
+    "from pathlib import Path\n",
+    "import tempfile\n",
+    "\n",
+    "from steerability.algorithms.core.steering_pipeline import SteeringPipeline\n",
+    "from steerability.algorithms.state_control.vjp_delta import VJPDelta\n",
+    "from steerability.spipe import SPipe\n",
+    "from tests.utils.tiny_models import tiny_llama, wordlevel_tokenizer\n",
```
[the remark this answers is not in the document]
Comment (wassname) [reply examples/notebooks/algorithms/vjp_delta.ipynb]: Replaced test fixtures with standard HF loading in 9ad0a46. — Pi

File: slop/verification/2026-09-17_vjp-delta/gpu_smoke_status.json [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  "created_at": "2026-09-17T18:23:42.566443786+08:00",
+  "original_command": "env PYTHONPATH=/home/code/dev/steerability-vjp /home/code/dev/steerability/.venv/bin/python slop/verification/2026-09-17_vjp-delta/gpu_smoke.py",
+  "command": "env PYTHONPATH=/home/code/dev/steerability-vjp /home/code/dev/steerability/.venv/bin/python slop/verification/2026-09-17_vjp-delta/gpu_smoke.py",
+  "path": "/home/code/dev/steerability-vjp",
+  "envs": {
+    "BUN_INSTALL": "/home/code/.bun",
+    "FZF_ALT_C_COMMAND": "command fd -H --no-ignore-vcs -E .git -td",
+    "forgit_clean": "gclean",
+    "forgit_switch_branch": "gsw",
+    "forgit_restore": "grs",
```
[the remark this answers is not in the document]
Comment (wassname) [reply slop/verification/2026-09-17_vjp-delta/gpu_smoke_status.json]: Removed all temporary artifacts from the PR diff in 9ad0a46. — Pi

File: steerability/algorithms/state_control/vjp_delta/fit.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+        if model is None:
+            raise ValueError("VJP-delta fitting requires a live model at steer time.")
+        if torch.is_inference_mode_enabled():
+            raise RuntimeError("VJP-delta fitting cannot run while torch.inference_mode() is enabled.")
+        target_layer, source_ids = self._layers(model)
+        layout = resolve_model_layout(model)
+        target_module = model.get_submodule(layout.layer_names[target_layer])
+        source_modules = {layer_id: model.get_submodule(layout.layer_names[layer_id]) for layer_id in source_ids}
+        try:
+            device = next(model.parameters()).device
```
[the remark this answers is not in the document]
Comment (wassname) [reply steerability/algorithms/state_control/vjp_delta/fit.py]: Added an explicit single-device check in 9ad0a46. — Pi

File: steerability/algorithms/state_control/vjp_delta/fit.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+                positive = self._class_gradients(
+                    model, tokenizer, self.data.positives, target_module=target_module,
+                    source_modules=source_modules, cotangent=cotangent, device=device,
+                )
+                negative = self._class_gradients(
+                    model, tokenizer, self.data.negatives, target_module=target_module,
+                    source_modules=source_modules, cotangent=cotangent, device=device,
+                )
+        finally:
+            model.train(was_training)
```
[the remark this answers is not in the document]
Comment (wassname) [reply steerability/algorithms/state_control/vjp_delta/fit.py]: Restores each module’s training flag in 9ad0a46. — Pi

File: steerability/algorithms/state_control/vjp_delta/control.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    uses the standard additive intervention and token scopes.
+
+    A precomputed `SteeringVector` or `ArtifactSource` avoids gradient extraction. The frozen form
+    is `ActivationAdapter`, so a reloaded `.spipe` resolves the stored vectors without a VJP fit.
+
+    Reference:
+
+    - VJP-delta method implementation provenance: `wassname/vjp-steering` at
+      `cb03382ebd0cc9cad615d169f42e68e8ae3e12a7`
+      [https://github.com/wassname/vjp-steering](https://github.com/wassname/vjp-steering)
```
[the remark this answers is not in the document]
Comment (wassname) [reply steerability/algorithms/state_control/vjp_delta/control.py:31]: Added the repo citation and J-lens reference in 9ad0a46. — Pi

File: steerability/algorithms/state_control/vjp_delta/fit.py [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+
+    data: LabeledExamples | dict
+    target_layer: int | None = None
+    source_layer_ids: Sequence[int] | None = None
+    skip_first: int = 16
+    max_length: int = 384
+    batch_size: int = 8
+
+    _model_ref: weakref.ref | None = field(default=None, init=False, repr=False, compare=False)
+    _master: SteeringVector | None = field(default=None, init=False, repr=False, compare=False)
```
[the remark this answers is not in the document]
Comment (wassname) [reply steerability/algorithms/state_control/vjp_delta/fit.py:99]: `spipe.codec._encode_dataclass` excludes `init=False` cache fields; verified identical digests before and after fitting. — Pi

File: steerability/algorithms/state_control/vjp_delta/args.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+from steerability.algorithms.state_control.common.steering_vector import SteeringVector
+from steerability.algorithms.state_control.common.token_scope import ScopeKind
+
+
+@dataclass
+class VJPDeltaArgs(BaseArgs):
+    """Arguments for `VJPDelta`.
+
+    Args:
+        steering_vector: A precomputed vector or artifact source. It skips gradient extraction.
```
[the remark this answers is not in the document]
Comment (wassname) [reply steerability/algorithms/state_control/vjp_delta/args.py]: Clarified precomputed vectors versus fitting sources in 18ad0ba. — Pi

File: steerability/algorithms/state_control/vjp_delta/control.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    """VJP-delta activation steering.
+
+    The control fits one normalized additive direction per source layer during `steer()`. Its
+    target contrast is the positive-minus-negative final unpadded target state. The fit applies
+    that contrast as a cotangent at valid target tokens and averages valid source-token gradients
+    per prompt before separately averaging the positive and negative classes. At generation it
+    uses the standard additive intervention and token scopes.
+
+    A precomputed `SteeringVector` or `ArtifactSource` avoids gradient extraction. The frozen form
+    is `ActivationAdapter`, so a reloaded `.spipe` resolves the stored vectors without a VJP fit.
```
[the remark this answers is not in the document]
Comment (wassname) [reply steerability/algorithms/state_control/vjp_delta/control.py]: Corrected the same wording in 18ad0ba. — Pi
